### PR TITLE
Tabular: Fixed LightGBM HPO

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

Major Changes:
- Significantly simplified lgb_trial.py to call model.fit.
- Corrected HPO model path reporting and names
- Enabled time-based early stopping of HPO trials for LightGBM
- Enabled memory-based early stopping of HPO trials for LightGBM
- Significantly simplified LightGBM callback code, merged two existing callbacks into one
- Corrected LightGBM HPO hyperparameters (Previously was throwing away seed and num_boost_rounds, + altering original model hyperparameters incorrectly)
- Optimized early stopping for LightGBM HPO to use dynamic early stopping rounds

Note: These issues still exist for the other models. Because this is a major change, I will leave the code changes for other models to separate PR's.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Innixma** [comment]
@razmikmelikbekyan This may be of interest to you, feel free to leave comments.

**Comment by szha** [comment]
Job PR-333-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-333/1/index.html

**Comment by szha** [comment]
Job PR-333-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-333/2/index.html

File: autogluon/utils/tabular/ml/models/lgb/callbacks.py [hunk 1]
```diff
@@ -142,14 +148,27 @@ def _callback(env):
             _init(env)
         if not enabled[0]:
             return
-
+        if train_loss_name is not None:
+            train_loss_evals = [eval for i, eval in enumerate(env.evaluation_result_list) if eval[0] == 'train_set' and eval[1] == train_loss_name]
```
Comment (razmikmelikbekyan) [inline autogluon/utils/tabular/ml/models/lgb/callbacks.py]: just small comment to old code: `enumerate` is not needed.

File: autogluon/utils/tabular/ml/models/lgb/lgb_model.py [hunk 2]
```diff
@@ -277,23 +264,27 @@ def hyperparameter_tune(self, X_train, X_test, Y_train, Y_test, scheduler_option
...
         logger.log(15, "Best hyperparameter configuration for Gradient Boosting Model: ")
         logger.log(15, best_hp)
-        return (hpo_models, hpo_model_performances, hpo_results)
-        # TODO: do final fit here?
-        # args.final_fit = True
-        # final_model = scheduler.run_with_config(best_config)
-        # save(final_model)
+        return hpo_models, hpo_model_performances, hpo_results
+
+    def _get_train_loss_name(self):
```
Comment (razmikmelikbekyan) [inline autogluon/utils/tabular/ml/models/lgb/lgb_model.py]: Here I have comment question. As I understand it is for early stopping. Why we are not using the same loss which is used for `fit`  ?

Comment (Innixma) [reply autogluon/utils/tabular/ml/models/lgb/lgb_model.py]: This is not for early stopping actually, this is just recorded in reporter for info to the user when they call fit_summary() if I'm not mistaken, although that code was not written by me. The loss metric is often not equivalent to the evaluation metric, since an evaluation metric can often not be directly optimized for. Recording the training evaluation metric would make sense, although we currently don't do it. I personally don't see much benefit in general from observing training metrics, although @jwmueller would know the intent best.

Comment (razmikmelikbekyan) [reply autogluon/utils/tabular/ml/models/lgb/lgb_model.py]: Thanks for the explanation, I got the purpose of doing it. But still I have question. Let me reformulate it.  What will happen if for regression I put `eval_metric="r2"` when I run `fit()` ? 
I tried to debug a bit and as far as I see that in this line https://github.com/awslabs/autogluon/blob/b9fb6671c1ab70ddd29e931ef163bee512a86e8c/autogluon/utils/tabular/ml/models/lgb/hyperparameters/lgb_trial.py#L77

the `train_params['params']['objective']` is equal to `regression` which is equivalent to `l2` loss according to LightGBM documentation (https://lightgbm.readthedocs.io/en/latest/Parameters.html#objective). Which means we run `lgb.train()` with loss `l2` rather than `r2`. Is this what we want to do ? So this means that regardless which `eval_metric` I will use for regression problem, we always train the model with `l2` loss. This will work since minimising one loss should minimise the other also, but want to be sure that it is the behaviour which went to have.

Comment (jwmueller) [reply autogluon/utils/tabular/ml/models/lgb/lgb_model.py]: Yes this code is just to track the performance of the model at each round of iterative training, so the user can inspect the training runs later.

Re the objective mismatch: not every 3rd party model supports all the objectives that autogluon has, so we sometimes will have to make approximations.  However, we are fortunate in this case: minimizing `l2` is equivalent to maximizing `r2`.  This approximation would be less appropriate if we were doing something like quantile regression (say `MAE` loss to estimate median)  and telling `lgb.train()` to still minimize `l2` loss.  So we'd love to have your help if you notice any metric mismatches between what the user specified to autogluon in `eval_metric` and what we pass to each of the models as their objective.

File: autogluon/utils/tabular/ml/models/lgb/callbacks.py (hunk 1 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/models/lgb/callbacks.py]: Thanks! Removed in latest update.

**Comment by szha** [comment]
Job PR-333-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-333/3/index.html

File: autogluon/utils/tabular/ml/models/abstract/abstract_model.py [hunk 3]
```diff
@@ -314,7 +314,7 @@ def convert_to_compressed_template(self):
         template.path = template.create_contexts(self.path + template.name + os.path.sep)
         return template
 
-    def hyperparameter_tune(self, X_train, X_test, Y_train, Y_test, scheduler_options=None, **kwargs):
+    def hyperparameter_tune(self, X_train, X_test, Y_train, Y_test, scheduler_options, **kwargs):
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/abstract/abstract_model.py]: should every model's hyperparameter_tune require scheduler_options, when some don't even utilize a scheduler?

Comment (Innixma) [reply autogluon/utils/tabular/ml/models/abstract/abstract_model.py]: For the ones that don't they can have it default to None, but for AbstractModel's version it is required:

```
scheduler_func, scheduler_options = scheduler_options  # Unpack tuple
        if scheduler_func is None or scheduler_options is None:
            raise ValueError("scheduler_func and scheduler_options cannot be None for hyperparameter tuning")
```
